python_modules/dagster-graphql/dagster_graphql/implementation/execution/__init__.py

```diff
@@ -91,21 +96,27 @@ def terminate_pipeline_execution(
         terminate_policy == GrapheneTerminateRunPolicy.MARK_AS_CANCELED_IMMEDIATELY
     )
 
-    if not record:
-        assert_permission(graphene_info, Permissions.TERMINATE_PIPELINE_EXECUTION)
-        return GrapheneRunNotFoundError(run_id)
+    try:
+        if not record:
+            assert_permission(graphene_info, Permissions.TERMINATE_PIPELINE_EXECUTION)
+            return GrapheneRunNotFoundError(run_id)
 
-    run = record.dagster_run
-    graphene_run = GrapheneRun(record)
+        run = record.dagster_run
+        graphene_run = GrapheneRun(record)
 
-    location_name = run.external_job_origin.location_name if run.external_job_origin else None
+        location_name = run.external_job_origin.location_name if run.external_job_origin else None
 
-    if location_name:
-        assert_permission_for_location(
-            graphene_info, Permissions.TERMINATE_PIPELINE_EXECUTION, location_name
+        if location_name:
+            assert_permission_for_location(
+                graphene_info, Permissions.TERMINATE_PIPELINE_EXECUTION, location_name
+            )
+        else:
+            assert_permission(graphene_info, Permissions.TERMINATE_PIPELINE_EXECUTION)
+
+    except UserFacingGraphQLError as e:
+        return GrapheneTerminateRunUnauthorizedError(
+            run_id=run_id, message=e.error.message if e.error.message else "Authorization failed"
         )
```

Comment:
i don't think its necc correct to assume that any UserFacingGraphQLError is a permissions error (plus i think its good to avoid using exceptions as control flow)

What i would suggest is removing the try/catch, and instead of calling assert_permission and assert_permission_for_location and potentially raising an exception, check has_permission and has_permission_for_location, and return the value you want if they fail

I think this could return a GrapheneTerminateRunFailure instead of a new type that the frontend needs to worry about - do we need to worry at all about this affecting the frontend callsite for the single run termination graphql call?